dynamic and nint decoding support

333fred opened the pull request:
Supports decoding dynamic and nint nested elements of a function pointer type.

Relates to https://github.com/dotnet/roslyn/issues/43321 (test plan for function pointers)

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for review.

jcouv commented:
>                 // Encode transforms flag for this type and it's custom modifiers (if any).

not related to this PR: typo

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:865 in 5c72cd1. [](commit_id = 5c72cd16aa895c92a189f5ae00adc6729bd96c29, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +913,42 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+                        handle(param.RefKind, param.RefCustomModifiers, param.TypeWithAnnotations);
+                    }
+
+                    void handle(RefKind refKind, ImmutableArray<CustomModifier> customModifiers, TypeWithAnnotations twa)
```
>handle [](start = 25, length = 6)

nit: we've tried to avoid nested local function, for readability. If moved up, could rename to `handleFunctionPointerParameter`
nit: `twa` -> `type`, `customModifiers` -> `refCustomModifiers`

333fred replied:
It cannot, it captures things from the current scope.

---
In reply to: [422327380](https://github.com/dotnet/roslyn/pull/44078#discussion_r422327380) [](ancestors = 422327380)

jcouv replied:
It doesn't have to capture things from the current scope :-)

---
In reply to: [422416731](https://github.com/dotnet/roslyn/pull/44078#discussion_r422416731) [](ancestors = 422416731,422327380)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +913,42 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
                 // Continue walking types
                 return false;
+
+                static void handleFunctionPointerType(FunctionPointerTypeSymbol funcPtr, ArrayBuilder<bool> transformFlagsBuilder, bool isNestedNamedType, bool addCustomModifierFlags)
```
>isNestedNamedType [](start = 136, length = 17)

Do we have a test involving nested named types in a function pointer?

333fred replied:
We cannot, but it does mean I need to find a test that breaks the bug I have.

---
In reply to: [422328571](https://github.com/dotnet/roslyn/pull/44078#discussion_r422328571) [](ancestors = 422328571)

AlekseyTs replied:
> We cannot, but it does mean I need to find a test that breaks the bug I have.

The parameter looks unused to me in iteration 4

---
In reply to: [422419465](https://github.com/dotnet/roslyn/pull/44078#discussion_r422419465) [](ancestors = 422419465,422328571)

333fred replied:
Exactly. There's a bug, I need to find a test.

---
In reply to: [422530525](https://github.com/dotnet/roslyn/pull/44078#discussion_r422530525) [](ancestors = 422530525,422419465,422328571)

333fred replied:
Talking with Aleksey, this should never be true for function pointer types. I'll assert it's false.

---
In reply to: [423322684](https://github.com/dotnet/roslyn/pull/44078#discussion_r423322684) [](ancestors = 423322684,422530525,422419465,422328571)

AlekseyTs replied:
> Talking with Aleksey, this should never be true for function pointer types. I'll assert it's false.

I suggest asserting at the call site and removing the parameter here.

---
In reply to: [423336064](https://github.com/dotnet/roslyn/pull/44078#discussion_r423336064) [](ancestors = 423336064,423322684,422530525,422419465,422328571)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:
```diff
@@ -696,33 +696,37 @@ private static bool IsTypeLessVisibleThan(TypeSymbol type, Symbol sym, ref HashS
-                var result = visitType(currentPointer.ReturnTypeWithAnnotations);
+                var result = VisitType(
+                    typeWithAnnotationsOpt: canDigThroughNullable ? default : currentPointer.ReturnTypeWithAnnotations,
+                    type: canDigThroughNullable ? currentPointer.ReturnTypeWithAnnotations.NullableUnderlyingTypeOrSelf : null,
```
>currentPointer.ReturnTypeWithAnnotations [](start = 50, length = 40)

consider extracting local. Same for `parameter.TypeWithAnnotations`

It looks like this is just inlining the old code. Or did I miss some difference?

333fred replied:
This is feedback from the previous PR.

---
In reply to: [422348522](https://github.com/dotnet/roslyn/pull/44078#discussion_r422348522) [](ancestors = 422348522)

jcouv replied:
So it is just inlining, or there is some difference? I'm not casting judgement, I've starred at the diff and want to make sure I understood :-)

---
In reply to: [422419109](https://github.com/dotnet/roslyn/pull/44078#discussion_r422419109) [](ancestors = 422419109,422348522)

AlekseyTs replied:
> So it is just inlining, 

The motivation is to reduce used stack size by eliminating a capture and unnecessary stack frames.

---
In reply to: [422422459](https://github.com/dotnet/roslyn/pull/44078#discussion_r422422459) [](ancestors = 422422459,422419109,422348522)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -338,6 +341,78 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol pointerType)
+                return type;
+            }
+
+            static TypeSymbol? handle(ref DynamicTypeDecoder decoder, RefKind refKind, ImmutableArray<CustomModifier> customModifiers, TypeWithAnnotations twa)
```
>customModifiers [](start = 118, length = 15)

Consider `customModifiers` -> `refCustomModifiers`. Also `twa` -> `type`

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -338,6 +341,78 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol pointerType)
+
+            static TypeSymbol? handle(ref DynamicTypeDecoder decoder, RefKind refKind, ImmutableArray<CustomModifier> customModifiers, TypeWithAnnotations twa)
+            {
+                if ((refKind != RefKind.None && decoder.ConsumeFlag())
```
>ConsumeFlag [](start = 56, length = 11)

Should we check it was a `false` flag?

jcouv replied:
>refKind != RefKind.None && decoder.ConsumeFlag() [](start = 21, length = 48)

consider `decoder.HandleParameterRefKind(refKind)`

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -338,6 +341,78 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol pointerType)
+            if (paramsTransformed || !sig.ReturnType.Equals(transformedReturn, TypeCompareKind.ConsiderEverything))
+            {
+                return type.SubstituteTypeSymbol(sig.ReturnTypeWithAnnotations.WithType(transformedReturn), transformedParameters,
+                                                 refCustomModifiers: default, paramRefCustomModifiers: default);
```
>refCustomModifiers: default, paramRefCustomModifiers: default [](start = 49, length = 61)

Are we intentionally dropping modifiers?

333fred replied:
No, passing `default` means that the existing mods, if they exist, will be passed.

---
In reply to: [422356877](https://github.com/dotnet/roslyn/pull/44078#discussion_r422356877) [](ancestors = 422356877)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/NativeIntegerTypeDecoder.cs:
```diff
@@ -137,6 +139,45 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol type)
+                foreach (var param in type.Signature.Parameters)
+                {
+                    var transformedParam = TransformTypeWithAnnotations(param.TypeWithAnnotations);
+                    paramsModified = paramsModified || !transformedParam.IsSameAs(param.TypeWithAnnotations);
```
>= paramsModified || [](start = 35, length = 19)

nit: `||=`?

333fred replied:
[https://github.com/dotnet/csharplang/issues/1718](https://github.com/dotnet/csharplang/issues/1718) if you want to try and convince the LDM to bring the idea back :wink:

---
In reply to: [422360892](https://github.com/dotnet/roslyn/pull/44078#discussion_r422360892) [](ancestors = 422360892)

jcouv commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -43,6 +43,17 @@ public static int CustomModifierCount(this TypeSymbol type)
+                case SymbolKind.FunctionPointer:
+                    {
+                        var funcPtr = (FunctionPointerTypeSymbol)type;
+                        var count = funcPtr.Signature.ReturnTypeWithAnnotations.CustomModifiers.Length + funcPtr.Signature.ReturnType.CustomModifierCount();
```
>funcPtr.Signature.ReturnTypeWithAnnotations.CustomModifiers.Length + funcPtr.Signature.ReturnType.CustomModifierCount() [](start = 36, length = 119)

I would have expected `funcPtr.Signature.ReturnTypeWithAnnotations.CustomModifiers` to be the same as `funcPtr.Signature.ReturnType.CustomModifierCount()`, so I feel we may be double-counting. Is that not the case?

jcouv commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -133,6 +161,8 @@ public static bool HasCustomModifiers(this TypeSymbol type, bool flagNonDefaultA
 
             return false;
+
+            bool checkTypeWithAnnotations(TypeWithAnnotations twa) => twa.CustomModifiers.Any() || twa.Type.HasCustomModifiers(flagNonDefaultArraySizesOrLowerBounds);
```
>checkTypeWithAnnotations [](start = 17, length = 24)

nit: consider `static` and passing in `flagNonDefaultArraySizesOrLowerBounds` for clarity

jcouv commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -96,14 +107,31 @@ public static bool HasCustomModifiers(this TypeSymbol type, bool flagNonDefaultA
-                        return pointedAtType.CustomModifiers.Any() || pointedAtType.Type.HasCustomModifiers(flagNonDefaultArraySizesOrLowerBounds);
+                        return checkTypeWithAnnotations(pointedAtType);
+                    }
+                case SymbolKind.FunctionPointer:
```
>case SymbolKind.FunctionPointer: [](start = 16, length = 32)

should we flag ref custom modifiers too?

cston commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +913,42 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+
+                static void handleFunctionPointerType(FunctionPointerTypeSymbol funcPtr, ArrayBuilder<bool> transformFlagsBuilder, bool isNestedNamedType, bool addCustomModifierFlags)
+                {
+                    Func<TypeSymbol, (ArrayBuilder<bool>, bool), bool, bool> visitor = addFlags;
```
>visitor = addFlags [](start = 77, length = 18)

This includes an allocation. And I think it would be clearer to inline the local and the local function anyway:
```
twa.Type.VisitType((type, param, isNestedNamedType) =>
    AddFlags(type, param.builder, isNestedNamedType, param.addCustomModifiers),
    (transformFlagsBuilder, addCustomModifierFlags));
```

333fred replied:
`addFlags` is static, so it should be cached. Personally, I don't believe that it's clearer.

---
In reply to: [422377182](https://github.com/dotnet/roslyn/pull/44078#discussion_r422377182) [](ancestors = 422377182)

jcouv replied:
I don't think there is caching here, even if the local function is static:
[example](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQZgAJ1CBhQgb10JuKJQBZCBZACgEpLraeUM1CAD0IBeQgBsA9gGMI4gNzceNJcsEdFOZbXXtN2mnwBsxJlNniOlQgF9VdnDaA==)

To cache, we would use a static Func field. Here's an example from `TypeSymbolExtensions`:
```
        private static readonly Func<TypeSymbol, TypeParameterSymbol?, bool, bool> s_containsTypeParameterPredicate =
            (type, parameter, unused) => type.TypeKind == TypeKind.TypeParameter && (parameter is null || TypeSymbol.Equals(type, parameter, TypeCompareKind.ConsiderEverything2));

```

---
In reply to: [422416920](https://github.com/dotnet/roslyn/pull/44078#discussion_r422416920) [](ancestors = 422416920,422377182)

cston replied:
We could use a `static` field. In this case, there is only one use of the `Func` though so it would be simpler to use a lambda which would be cached automatically.

---
In reply to: [422424873](https://github.com/dotnet/roslyn/pull/44078#discussion_r422424873) [](ancestors = 422424873,422416920,422377182)

cston commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -43,6 +43,17 @@ public static int CustomModifierCount(this TypeSymbol type)
+                        var count = funcPtr.Signature.ReturnTypeWithAnnotations.CustomModifiers.Length + funcPtr.Signature.ReturnType.CustomModifierCount();
+                        foreach (var param in funcPtr.Signature.Parameters)
+                        {
+                            count += param.TypeWithAnnotations.CustomModifiers.Length + param.Type.CustomModifierCount();
```
>param.TypeWithAnnotations.CustomModifiers.Length + param.Type.CustomModifierCount() [](start = 37, length = 83)

We have this same pattern in each of the cases. Please consider extracting a `static int getCustomModifierCount(TypeWithAnnotations)` method.

333fred replied:
I'll just bring the next commit from my planned follow-up pr forward which does more here.

---
In reply to: [422381885](https://github.com/dotnet/roslyn/pull/44078#discussion_r422381885) [](ancestors = 422381885)

cston commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -96,14 +107,31 @@ public static bool HasCustomModifiers(this TypeSymbol type, bool flagNonDefaultA
+                            }
+                        }
+
+                        return false;
```
Consider using:
```
return funcPtr.Signature.Parameters.Any(
    (p, f) => checkTypeWithAnnotations(p, f),
    flagNonDefaultArraySizesOrLowerBounds);
```

333fred replied:
These methods can be called on very deeply nested generics. I'd prefer to avoid any kind of capturing.

---
In reply to: [422382953](https://github.com/dotnet/roslyn/pull/44078#discussion_r422382953) [](ancestors = 422382953)

cston replied:
I don't think we should capture any state either. `checkTypeWithAnnotations` should be `static`.

---
In reply to: [423324444](https://github.com/dotnet/roslyn/pull/44078#discussion_r423324444) [](ancestors = 423324444,422382953)

333fred replied:
This needs to check both the TypeWithAnnotations and the ref custom modifiers, and I'm not planning on refactoring existing code beyond what I have to.

---
In reply to: [423326618](https://github.com/dotnet/roslyn/pull/44078#discussion_r423326618) [](ancestors = 423326618,423324444,422382953)

cston commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -119,7 +147,7 @@ public static bool HasCustomModifiers(this TypeSymbol type, bool flagNonDefaultA
 
                                 foreach (TypeWithAnnotations typeArg in typeArgs)
```
>foreach (TypeWithAnnotations typeArg in typeArgs) [](start = 32, length = 49)

Consider using:
```
if (typeArgs.Any(
    (p, f) => checkTypeWithAnnotations(p, f),
    flagNonDefaultArraySizesOrLowerBounds))
```

333fred replied:
These methods can be called on very deeply nested generics. I'd prefer to avoid any kind of capturing.

---
In reply to: [422383386](https://github.com/dotnet/roslyn/pull/44078#discussion_r422383386) [](ancestors = 422383386)

cston commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs:
```diff
@@ -1022,6 +1022,52 @@ void assert(string expectedType, string fieldName)
+}
+", options: TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.RegularPreview, symbolValidator: symbolValidator);
+
+            void symbolValidator(ModuleSymbol module)
```
>void [](start = 12, length = 4)

`static`

cston commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs:
```diff
@@ -1022,6 +1022,52 @@ void assert(string expectedType, string fieldName)
+                assert("delegate*<delegate*<System.IntPtr, System.IntPtr, System.IntPtr>, nint>", "F5");
+                assert("delegate*<nint, delegate*<System.IntPtr, System.IntPtr, System.IntPtr>>", "F6");
+
+                void assert(string expectedType, string fieldName)
```
>void [](start = 16, length = 4)

`static`

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5174,6 +5175,225 @@ void symbolValidator(ModuleSymbol module)
+
+                assertField("F14", "System.Runtime.CompilerServices.DynamicAttribute({false, false, false, false, true})", "delegate*<out modreq(System.Runtime.InteropServices.OutAttribute) dynamic, System.Object>");
+
+                void assertField(string field, string? expectedAttribute, string expectedType)
```
>assertField [](start = 21, length = 11)

nit: same comment on nested local functions

333fred replied:
The whole point of this local function is capturing things to make it easier to read and type.

---
In reply to: [422430322](https://github.com/dotnet/roslyn/pull/44078#discussion_r422430322) [](ancestors = 422430322)

jcouv commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs:
```diff
@@ -1022,6 +1022,52 @@ void assert(string expectedType, string fieldName)
+            {
+                var expectedAttributes = @"
+C
+    [NativeInteger({ False, True, True, True })] delegate*<System.IntPtr, System.IntPtr, System.IntPtr> F1
```
>False, True, True, True [](start = 21, length = 23)

Filed [https://github.com/dotnet/roslyn/issues/44086](https://github.com/dotnet/roslyn/issues/44086) for native int encoding question. Not sure why we would want to encode a bool for a type that is not an U/IntPtr.

jcouv commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NativeInteger.cs:
```diff
@@ -1022,6 +1022,52 @@ void assert(string expectedType, string fieldName)
+    public delegate*<System.IntPtr, System.IntPtr, nint> F2;
+    public delegate*<nint, System.IntPtr, System.IntPtr> F3;
+    public delegate*<System.IntPtr, nint, System.IntPtr> F4;
+    public delegate*<delegate*<System.IntPtr, System.IntPtr, System.IntPtr>, nint> F5;
```
>F5 [](start = 83, length = 2)

Consider adding a case where the `nint` is in the nested function pointer, rather than in the outer one

jcouv commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -280,8 +280,26 @@ public override void VisitMethod(IMethodSymbol symbol)
                     AddSpace();
                 }
 
+                if (ShouldMethodDisplayReadOnly(symbol))
```
>if (ShouldMethodDisplayReadOnly(symbol)) [](start = 16, length = 40)

Please add a test for `ToDisplayString`  that show the conditional behavior for the different parts depending on options.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5174,6 +5175,225 @@ void symbolValidator(ModuleSymbol module)
+
+                assertField("F1", "System.Runtime.CompilerServices.DynamicAttribute({false, true, true, true})", "delegate*<dynamic, dynamic, dynamic>");
+
+                assertField("F2", "System.Runtime.CompilerServices.DynamicAttribute({false, true, false, false})", "delegate*<System.Object, System.Object, dynamic>");
```
>true, false, false [](start = 92, length = 18)

Feels strange that the `true` comes first, despite corresponding with the last parameter. Am I reading this wrong?
Similarly, on F4, I don't understand the order of the flags (I would expect a `true` between two `false` flags)

333fred replied:
These are encoded in metadata order. Return parameter first, then actual parameters.

---
In reply to: [422433647](https://github.com/dotnet/roslyn/pull/44078#discussion_r422433647) [](ancestors = 422433647)

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5174,6 +5175,225 @@ void symbolValidator(ModuleSymbol module)
+  .method public hidebysig specialname rtspecialname 
+    instance void .ctor () cil managed 
+  {
+    .maxstack 8
```
>.maxstack 8 [](start = 4, length = 11)

nit: can be removed

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5174,6 +5175,225 @@ void symbolValidator(ModuleSymbol module)
+  .method public hidebysig static void TooManyFlags(method class [mscorlib]System.Object *(class [mscorlib]System.Object) a)
+  {
+    .param [1]
+    .custom instance void [mscorlib]System.Runtime.CompilerServices.DynamicAttribute::.ctor(bool[]) = ( 01 00 04 00 00 00 00 01 01 01 00 00 ) 
```
>( 01 00 04 00 00 00 00 01 01 01 00 00 )  [](start = 102, length = 40)

a comment decoding these as bool[] would be very useful ;-)

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -280,8 +280,26 @@ public override void VisitMethod(IMethodSymbol symbol)
                     AddSpace();
                 }
 
+                if (ShouldMethodDisplayReadOnly(symbol))
```
>ShouldMethodDisplayReadOnly [](start = 20, length = 27)

Is this applicable to a function pointer?

333fred replied:
No, I meant to remove it.

---
In reply to: [422525960](https://github.com/dotnet/roslyn/pull/44078#discussion_r422525960) [](ancestors = 422525960)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -887,6 +887,13 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+
+                        // Function pointer types have nested custom modifiers and refkinds in line with types, and visit all their nested types
+                        // as part of this call.
+                        return true;
```
>return true; [](start = 24, length = 12)

Could returning true cut the visit process short and leave some constituent parts unvisited?

333fred replied:
Yeah, this is busted for things like `D<delegate*<dynamic>[], dynamic>`. I'll need to think on this some more. Fundamentally we need to visit more than `VisitTypes` normally will visit here, and in a different order, but we also want to keep going after this type is passed.

---
In reply to: [422527325](https://github.com/dotnet/roslyn/pull/44078#discussion_r422527325) [](ancestors = 422527325)

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/44160](https://github.com/dotnet/roslyn/issues/44160) to track. This is going to be too big a change to handle in this PR, and I think this is a niche enough case (and only results in dynamic not being treated as dynamic) that I'd prefer to address later.

---
In reply to: [423335853](https://github.com/dotnet/roslyn/pull/44078#discussion_r423335853) [](ancestors = 423335853,422527325)

AlekseyTs commented on src/Compilers/CSharp/Portable/Utilities/TypeSymbolExtensions.cs:
```diff
@@ -43,6 +43,17 @@ public static int CustomModifierCount(this TypeSymbol type)
+                            count += param.TypeWithAnnotations.CustomModifiers.Length + param.Type.CustomModifierCount();
+                        }
+
+                        return count;
```
>count [](start = 31, length = 5)

We already have a helper below. Could we just call it here?
```
        /// <summary>
        /// Count the number of custom modifiers in/on the return type
        /// and parameters of the specified method.
        /// </summary>
        public static int CustomModifierCount(this MethodSymbol method)
```

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +913,42 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+                {
+                    Func<TypeSymbol, (ArrayBuilder<bool>, bool), bool, bool> visitor = addFlags;
+
+                    transformFlagsBuilder.Add(false);
```
>transformFlagsBuilder.Add(false); [](start = 20, length = 33)

I assume this is for the pointer type itself. Consider adding a comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +913,42 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+
+                        if (addCustomModifierFlags)
+                        {
+                            HandleCustomModifiers(customModifiers.Length, transformFlagsBuilder);
```
>customModifiers [](start = 50, length = 15)

These come before the ref. Even though, given the current implementation of HandleCustomModifiers the result is the same, consider handling things in metadata order.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -338,6 +341,78 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol pointerType)
+
+            if (paramsTransformed || !sig.ReturnType.Equals(transformedReturn, TypeCompareKind.ConsiderEverything))
+            {
+                return type.SubstituteTypeSymbol(sig.ReturnTypeWithAnnotations.WithType(transformedReturn), transformedParameters,
```
>sig.ReturnTypeWithAnnotations.WithType(transformedReturn) [](start = 49, length = 57)

t looks like this could be unnecessary if the type didn't change.

AlekseyTs commented:
Done with review pass (iteration 4), tests are not looked at.

333fred commented:
@AlekseyTs @cston @jcouv addressed feedback.

---
In reply to: [626221533](https://github.com/dotnet/roslyn/pull/44078#issuecomment-626221533) [](ancestors = 626221533)

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -338,6 +341,92 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol pointerType)
+            else
+            {
+                transformedReturnWithAnnotations = sig.ReturnTypeWithAnnotations;
+            }
```
Can `handle` transform `TypeWithAnnotations` instead of `TypeSymbol` to avoid duplicating this here and for parameters?

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -349,25 +349,12 @@ private PointerTypeSymbol TransformPointerType(PointerTypeSymbol pointerType)
-            var transformedReturn = handle(ref this, sig.RefKind, sig.RefCustomModifiers, sig.ReturnTypeWithAnnotations);
-            if (transformedReturn is null)
+            var (transformedReturnWithAnnotations, madeChanges) = handle(ref this, sig.RefKind, sig.RefCustomModifiers, sig.ReturnTypeWithAnnotations);
+            if (transformedReturnWithAnnotations is null)
```
`handle()` could return `TypeWithAnnotations` rather than `TypeWithAnnotations?` and the caller could use `if (transformedReturnWithAnnotations.IsDefault)`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +916,43 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+                static void handleFunctionPointerType(FunctionPointerTypeSymbol funcPtr, ArrayBuilder<bool> transformFlagsBuilder, bool addCustomModifierFlags)
+                {
+                    Func<TypeSymbol, (ArrayBuilder<bool>, bool), bool, bool> visitor =
+                        (TypeSymbol type, (ArrayBuilder<bool> builder, bool addCustomModiferFlags) param, bool _) => AddFlags(type, param.builder, isNestedNamedType: false, param.addCustomModiferFlags);
```
>bool _ [](start = 106, length = 6)

It feels inappropriate to ignore this parameter.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs:
```diff
@@ -906,6 +916,43 @@ private static bool AddFlags(TypeSymbol type, ArrayBuilder<bool> transformFlagsB
+                static void handleFunctionPointerType(FunctionPointerTypeSymbol funcPtr, ArrayBuilder<bool> transformFlagsBuilder, bool addCustomModifierFlags)
+                {
+                    Func<TypeSymbol, (ArrayBuilder<bool>, bool), bool, bool> visitor =
+                        (TypeSymbol type, (ArrayBuilder<bool> builder, bool addCustomModiferFlags) param, bool _) => AddFlags(type, param.builder, isNestedNamedType: false, param.addCustomModiferFlags);
```
>false [](start = 166, length = 5)

Always passing false doesn't feel right.

AlekseyTs replied:
It doesn't look like this fix is covered by a test.

---
In reply to: [424088267](https://github.com/dotnet/roslyn/pull/44078#discussion_r424088267) [](ancestors = 424088267)

333fred replied:
@AlekseyTs test added.

---
In reply to: [424722117](https://github.com/dotnet/roslyn/pull/44078#discussion_r424722117) [](ancestors = 424722117,424088267)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs:
```diff
@@ -201,7 +204,7 @@ private bool HandleCustomModifiers(int customModifiersCount)
         }
 
         // Native compiler encodes bools (always false) for custom modifiers and parameter ref-kinds, if ref-kind is ref or out.
```
>parameter ref-kinds, if ref-kind is ref or out. [](start = 81, length = 47)

Consider adjusting the comment as well.

AlekseyTs commented:
Done with review pass (iteration 8), tests are not looked at.

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [627661089](https://github.com/dotnet/roslyn/pull/44078#issuecomment-627661089) [](ancestors = 627661089)
